File: Engine/Android/core/src/test/kotlin/despia/engine/LinkPreviewConformanceTest.kt
package despia.engine

import java.io.File
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import org.junit.jupiter.api.DynamicTest
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestFactory

/**
 * The SHARED LINK-PREVIEW corpus (OpenSource/Conformance/preview/metadata.json) through THIS
 * runtime's LinkPreview core — the Kotlin leg of Core/Preview (F17.1). The TS reference
 * (packages/kernel/test/linkpreview-conformance.test.ts) and the Swift twin
 * (Engine/iOS/LinkPreview.swift) read the SAME file, so `og:image` cannot resolve to three
 * different URLs on three renderers.
 */
class LinkPreviewConformanceTest {

    private fun corpusDir(): File {
        var dir = File(System.getProperty("user.dir") ?: ".").absoluteFile
        while (true) {
            val candidate = File(dir, "OpenSource/Conformance/preview")
            if (candidate.isDirectory) return candidate
            dir = dir.parentFile ?: error("OpenSource/Conformance/preview not found")
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun doc(): Map<String, Any?> =
        json(File(corpusDir(), "metadata.json").readText()).foundationValue as? Map<String, Any?>
            ?: error("metadata.json: not a JSON object")

    @Suppress("UNCHECKED_CAST")
    private fun rows(key: String, minimum: Int): List<Map<String, Any?>> {
        val list = doc()[key] as? List<Map<String, Any?>> ?: error("no $key[]")
        assertTrue(list.size >= minimum, "preview/$key is suspiciously small (${list.size})")
        return list
    }

    private fun str(v: Any?): String = v as? String ?: ""
    private fun name(c: Map<String, Any?>): String = str(c["name"]).ifEmpty { "?" }

    private fun targetOf(raw: String): LinkPreview.Target =
        LinkPreview.resolveTarget(raw).getOrElse { error("corpus base is not a valid URL: $raw") }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun budget() {
        val b = doc()["budget"] as Map<String, Any?>
        assertEquals((b["maxBytes"] as Number).toInt(), LinkPreview.MAX_BYTES, "maxBytes")
        assertEquals((b["maxRedirects"] as Number).toInt(), LinkPreview.MAX_REDIRECTS, "maxRedirects")
        assertEquals((b["defaultTimeoutMs"] as Number).toInt(), LinkPreview.DEFAULT_TIMEOUT_MS, "defaultTimeoutMs")
        assertEquals((b["minTimeoutMs"] as Number).toInt(), LinkPreview.MIN_TIMEOUT_MS, "minTimeoutMs")
        assertEquals((b["maxTimeoutMs"] as Number).toInt(), LinkPreview.MAX_TIMEOUT_MS, "maxTimeoutMs")
    }

    @TestFactory
    fun timeout(): List<DynamicTest> = rows("timeout", 6).map { c ->
        DynamicTest.dynamicTest("preview/timeout — ${name(c)}") {
            assertEquals((c["expect"] as Number).toInt(), LinkPreview.clampTimeout(c["raw"]))
        }
    }

    @TestFactory
    @Suppress("UNCHECKED_CAST")
    fun url(): List<DynamicTest> = rows("url", 18).map { c ->
        DynamicTest.dynamicTest("preview/url — ${name(c)}") {
            val got = LinkPreview.resolveTarget(c["raw"])
            val expectOk = c["ok"] == true
            assertEquals(expectOk, got.isSuccess, "ok")
            if (expectOk) {
                val want = c["value"] as Map<String, Any?>
                val t = got.getOrThrow()
                assertEquals(str(want["url"]), t.url, "url")
                assertEquals(str(want["origin"]), t.origin, "origin")
                assertEquals(str(want["host"]), t.host, "host")
                assertEquals(str(want["scheme"]), t.scheme, "scheme")
                assertEquals(str(want["path"]), t.path, "path")
            } else {
                assertEquals(str(c["error"]), LinkPreview.code(got.exceptionOrNull()!!), "error")
            }
        }
    }

    @TestFactory
    fun reference(): List<DynamicTest> = rows("reference", 14).map { c ->
        DynamicTest.dynamicTest("preview/reference — ${name(c)}") {
            assertEquals(str(c["expect"]), LinkPreview.resolveReference(targetOf(str(c["base"])), c["ref"]))
        }
    }

    @TestFactory
    fun entities(): List<DynamicTest> = rows("entities", 6).map { c ->
        DynamicTest.dynamicTest("preview/entities — ${name(c)}") {
            assertEquals(str(c["expect"]), LinkPreview.decodeEntities(str(c["raw"])))
        }
    }

    @TestFactory
    fun collapse(): List<DynamicTest> = rows("collapse", 3).map { c ->
        DynamicTest.dynamicTest("preview/collapse — ${name(c)}") {
            assertEquals(str(c["expect"]), LinkPreview.collapseText(str(c["raw"])))
        }
    }

    @TestFactory
    fun siteFallback(): List<DynamicTest> = rows("siteFallback", 3).map { c ->
        DynamicTest.dynamicTest("preview/siteFallback — ${name(c)}") {
            assertEquals(str(c["expect"]), LinkPreview.siteFallback(str(c["host"])))
        }
    }

    @TestFactory
    @Suppress("UNCHECKED_CAST")
    fun parse(): List<DynamicTest> = rows("parse", 14).map { c ->
        DynamicTest.dynamicTest("preview/parse — ${name(c)}") {
            val got = LinkPreview.parse(str(c["html"]), targetOf(str(c["requestUrl"])))
            val want = c["expect"] as Map<String, Any?>
            assertEquals(str(want["title"]), got.title, "title")
            assertEquals(str(want["description"]), got.description, "description")
            assertEquals(str(want["image"]), got.image, "image")
            assertEquals(str(want["siteName"]), got.siteName, "siteName")
            assertEquals(str(want["favicon"]), got.favicon, "favicon")
            assertEquals(str(want["type"]), got.type, "type")
            assertEquals(str(want["canonical"]), got.canonical, "canonical")
        }
    }
}

File: Web/packages/kernel/test/linkpreview-conformance.test.ts
//
//  linkpreview-conformance.test.ts — the SHARED LINK-PREVIEW corpus
//  (OpenSource/Conformance/preview/metadata.json) through the TS core, the REFERENCE leg of
//  Core/Preview (F17.1). The Kotlin twin (:core LinkPreviewConformanceTest) and the Swift
//  twin (Engine/iOS/LinkPreview.swift, record lane) read the SAME file, so `og:image` cannot
//  resolve to three different URLs on three renderers.
//

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync, existsSync } from "node:fs";
import { join, dirname, resolve } from "node:path";

import {
  PREVIEW_MAX_BYTES, PREVIEW_MAX_REDIRECTS, PREVIEW_DEFAULT_TIMEOUT_MS,
  PREVIEW_MIN_TIMEOUT_MS, PREVIEW_MAX_TIMEOUT_MS,
  clampPreviewTimeout, resolvePreviewTarget, resolvePreviewReference,
  decodePreviewEntities, collapsePreviewText, previewSiteFallback, parseLinkPreview,
  type PreviewTarget,
} from "../src/linkpreview.ts";

function corpusDir(): string {
  let dir = resolve(import.meta.dirname ?? ".");
  for (;;) {
    const candidate = join(dir, "OpenSource/Conformance/preview");
    if (existsSync(candidate)) return candidate;
    const parent = dirname(dir);
    if (parent === dir) throw new Error("preview corpus not found");
    dir = parent;
  }
}

type Doc = {
  budget: { maxBytes: number; maxRedirects: number; defaultTimeoutMs: number; minTimeoutMs: number; maxTimeoutMs: number };
  timeout: { name: string; raw: unknown; expect: number }[];
  url: { name: string; raw: string; ok: boolean; value?: Record<string, string>; error?: string }[];
  reference: { name: string; base: string; ref: string; expect: string }[];
  entities: { name: string; raw: string; expect: string }[];
  collapse: { name: string; raw: string; expect: string }[];
  siteFallback: { name: string; host: string; expect: string }[];
  parse: { name: string; requestUrl: string; html: string; expect: Record<string, string> }[];
};

const doc = JSON.parse(readFileSync(join(corpusDir(), "metadata.json"), "utf-8")) as Doc;

assert.ok(doc.url.length >= 18, "preview/url corpus is suspiciously small");
assert.ok(doc.reference.length >= 14, "preview/reference corpus is suspiciously small");
assert.ok(doc.parse.length >= 14, "preview/parse corpus is suspiciously small");

test("preview — the pinned fetch budget", () => {
  assert.equal(PREVIEW_MAX_BYTES, doc.budget.maxBytes);
  assert.equal(PREVIEW_MAX_REDIRECTS, doc.budget.maxRedirects);
  assert.equal(PREVIEW_DEFAULT_TIMEOUT_MS, doc.budget.defaultTimeoutMs);
  assert.equal(PREVIEW_MIN_TIMEOUT_MS, doc.budget.minTimeoutMs);
  assert.equal(PREVIEW_MAX_TIMEOUT_MS, doc.budget.maxTimeoutMs);
});

for (const c of doc.timeout) {
  test(`preview/timeout — ${c.name}`, () => {
    assert.equal(clampPreviewTimeout(c.raw), c.expect);
  });
}

for (const c of doc.url) {
  test(`preview/url — ${c.name}`, () => {
    const got = resolvePreviewTarget(c.raw);
    assert.equal(got.ok, c.ok, "ok");
    if (got.ok) {
      assert.deepEqual(
        { url: got.value.url, origin: got.value.origin, host: got.value.host, scheme: got.value.scheme, path: got.value.path },
        c.value,
      );
    } else {
      assert.equal(got.error, c.error, "error");
    }
  });
}

function targetOf(raw: string): PreviewTarget {
  const got = resolvePreviewTarget(raw);
  assert.ok(got.ok, `corpus base is not a valid URL: ${raw}`);
  return got.value;
}

for (const c of doc.reference) {
  test(`preview/reference — ${c.name}`, () => {
    assert.equal(resolvePreviewReference(targetOf(c.base), c.ref), c.expect);
  });
}

for (const c of doc.entities) {
  test(`preview/entities — ${c.name}`, () => {
    assert.equal(decodePreviewEntities(c.raw), c.expect);
  });
}

for (const c of doc.collapse) {
  test(`preview/collapse — ${c.name}`, () => {
    assert.equal(collapsePreviewText(c.raw), c.expect);
  });
}

for (const c of doc.siteFallback) {
  test(`preview/siteFallback — ${c.name}`, () => {
    assert.equal(previewSiteFallback(c.host), c.expect);
  });
}

for (const c of doc.parse) {
  test(`preview/parse — ${c.name}`, () => {
    const got = parseLinkPreview(c.html, targetOf(c.requestUrl));
    assert.deepEqual(
      {
        title: got.title, description: got.description, image: got.image,
        siteName: got.siteName, favicon: got.favicon, type: got.type, canonical: got.canonical,
      },
      c.expect,
    );
  });
}
